src/librustc_mir/transform/qualify_min_const_fn.rs

```diff
@@ -0,0 +1,364 @@
+        tcx,
+        tcx.fn_sig(def_id).output().skip_binder(),
+        mir.local_decls.iter().next().unwrap().source_info.span,
+        token,
+    )?;
+
+    for bb in mir.basic_blocks() {
+        token = check_terminator(tcx, mir, bb.terminator(), token)?;
+        for stmt in &bb.statements {
+            token = check_statement(tcx, mir, stmt, token)?;
+        }
+    }
+    Ok(token)
+}
+
+fn check_ty(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    ty: ty::Ty<'tcx>,
+    span: Span,
+    token: IsMinConstFn,
+) -> McfResult {
+    for ty in ty.walk() {
+        match ty.sty {
+            ty::Anon(..) => return Err((span, "`impl Trait` in const fn is unstable".into())),
+            ty::FnPtr(..) => {
+                return Err((span, "function pointers in const fn are unstable".into()))
+            }
+            ty::Dynamic(preds, _) => {
+                for pred in preds.iter() {
+                    match pred.skip_binder() {
+                        | ty::ExistentialPredicate::AutoTrait(_)
+                        | ty::ExistentialPredicate::Projection(_) => {
+                            return Err((
+                                span,
+                                "trait bounds other than `Sized` \
+                                 on const fn parameters are unstable"
+                                    .into(),
+                            ))
+                        }
+                        ty::ExistentialPredicate::Trait(trait_ref) => {
+                            if Some(trait_ref.def_id) != tcx.lang_items().sized_trait() {
+                                return Err((
+                                    span,
+                                    "trait bounds other than `Sized` \
+                                     on const fn parameters are unstable"
+                                        .into(),
+                                ));
+                            }
+                        }
+                    }
+                }
+            }
+            _ => {}
+        }
+    }
+    Ok(token)
+}
+
+fn check_rvalue(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &'a Mir<'tcx>,
+    rvalue: &Rvalue<'tcx>,
+    span: Span,
+    token: IsMinConstFn,
+) -> McfResult {
+    match rvalue {
+        Rvalue::Repeat(operand, _) | Rvalue::Use(operand) => {
+            check_operand(tcx, mir, operand, span, token)
+        }
+        Rvalue::Len(place) | Rvalue::Discriminant(place) | Rvalue::Ref(_, _, place) => {
+            check_place(tcx, mir, place, span, token, PlaceMode::Read)
+        }
+        Rvalue::Cast(_, operand, cast_ty) => {
+            use rustc::ty::cast::CastTy;
+            let cast_in = CastTy::from_ty(operand.ty(mir, tcx)).expect("bad input type for cast");
+            let cast_out = CastTy::from_ty(cast_ty).expect("bad output type for cast");
+            match (cast_in, cast_out) {
+                (CastTy::Ptr(_), CastTy::Int(_)) | (CastTy::FnPtr, CastTy::Int(_)) => Err((
```

Comment:
And I guess this is where we hope that there are no other similar casts... (like `RPtr` to `Int`)